migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/migrate01/MigratorSchema.scala

```diff
@@ -0,0 +1,134 @@
+  private val log: Logger = LogManager.getLogger(this.getClass)
+
+  override protected val collectionBase: String = EntityMigrator.schemaCollection
+  override protected val collectionOld: String = MigrationUtils.getVersionedCollectionName(collectionBase, 0)
+  override protected val collectionNew: String = MigrationUtils.getVersionedCollectionName(collectionBase, 1)
+
+  protected val dbOld: MongoDatabase = databaseOld
+  protected val dbNew: MongoDatabase = databaseNew
+
+  /**
+    * Migrates a single instance of schema.
+    *
+    * @param srcSchemaJson A JSON representation of a schema in the model 0 format.
+    * @param objectId      An Object Id if the schema.
+    * @param repo          An entity repository.
+    */
+  def migrateEntity(srcSchemaJson: String, objectId: String, repo: EntityRepository): Unit = {
+    val schema1Opt = try {
+      val schema0 = Serializer0.deserializeSchema(srcSchemaJson)
+      Option(model1.Schema(
+        schema0.name,
+        schema0.version,
+        None,
+        userCreated = migrationUserName,
+        userUpdated = migrationUserName,
+        fields = schema0.fields.map(MigrationToV1.convertSchemaField(_, Nil))
+      ))
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Encountered a serialization error for '$collectionBase': ${e.getMessage}")
+        None
+    }
+
+    schema1Opt.foreach(schema1 => {
+      if (repo.doesDocumentExist(schema1.name, schema1.version)) {
+        resolveConflict(schema1, objectId, repo)
+      } else {
+        normalInsert(schema1, objectId, repo)
+      }
+    })
+  }
+
+  /**
+    * In case there is no conflict, insert a new schema normally.
+    * If inserting is unsuccessful fallback to conflict resolution.
+    *
+    * @param schema   A schema to save as an instance of Model 1 object
+    * @param objectId An Object Id of the schema
+    * @param repo     An entity repository
+    */
+  def normalInsert(schema: model1.Schema, objectId: String, repo: EntityRepository): Unit = {
+    val schema1Json = ObjectIdTools.putObjectIdIfNotPresent(Serializer1.serializeSchema(schema), Option(objectId))
+
+    try {
+      repo.insertDocument(schema1Json)
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Unable to append a document for 'schema': ${e.getMessage}. Retrying...")
+        resolveConflict(schema, objectId, repo)
+    }
+  }
+
+  /**
+    * In case there is a conflict, resolve it by adding a new version of the schema
+    * and register the mapping between the old version and the new one in ahe entity version map.
+    *
+    * @param schema   A schema to save as an instance of Model 1 object
+    * @param objectId An Object Id if the schema
+    * @param repo     An entity repository
+    */
+  def resolveConflict(schema: model1.Schema, objectId: String, repo: EntityRepository): Unit = {
+    var retriesLeft = EntityMigrator.NumberOfInsertRetries
+    var saved = false
```

Comment:
As always I don't like `var`s. If you have the will or time, we could eliminate them by having the `try` block in a different recursive method and using break instead of `saved`

Reply:
I'd prefer leaving it this way. The vars are confined in a method and the `while` loop approach seems easier to read for me.
If this still is a concert for you, please let me know, I'm going to rewrite it recursively.